shepherd_bms/src/main.cpp

```diff
@@ -233,32 +268,80 @@ void shepherdMain()
 		}
 	}
 
-	uint16_t packChargeVolt = 300;
-
 	// CHARGE STATE
 	if (digitalRead(CHARGE_DETECT) == LOW && bmsFault == FAULTS_CLEAR) {
-		digitalWrite(CHARGE_SAFETY_RELAY, HIGH);
-		compute.enableCharging(true);
-		if (currTime > lastChargeMsg + 150) {
+		// Check if we should charge
+		if (chargingCheck(accData)) {
+			digitalWrite(CHARGE_SAFETY_RELAY, HIGH);
+			compute.enableCharging(true);
+			compute.sendChargingStatus(true);
+		} else {
+			digitalWrite(CHARGE_SAFETY_RELAY, LOW);
+			compute.enableCharging(false);
+			compute.sendChargingStatus(false);
+		}
+
+		// Check if we should balance
+		if (balancingCheck(accData)) {
+			chargeBalancing(accData);
+		} else {
+			segment.enableBalancing(false);
+		}
+		
+		// Send CAN message, but not too often
+		if (currTime > lastChargeMsg + 250) {
 			lastChargeMsg = currTime;
-			compute.sendChargingMessage(packChargeVolt, accData->chargeLimit);
+			compute.sendChargingMessage(MAX_CHARGE_VOLT * NUM_CELLS_PER_CHIP * NUM_CHIPS, accData->chargeLimit, calcStateOfCharge(accData), accData);
```

Comment:
We should add SoC to the accumulatorData type and then just pass that, since SoC is also something that is a part of the BMS status sent out to the dashboard and other stuff

Reply:
lowkey we should just be passing accData for all these CAN messages